ui/src/app/views/home/heatmap/home.heatmap.component.ts

```diff
@@ -106,26 +121,79 @@ export class HomeHeatmapComponent implements OnInit {
+                            if (!this.unfilteredWorkflows[event.project_key]) {
+                                this.unfilteredWorkflows[event.project_key] = new Array<string>();
                             }
-                            if (this.workflows[event.project_key].lastIndexOf(event.workflow_name) === -1) {
-                                this.workflows[event.project_key].push(event.workflow_name);
+                            if (this.unfilteredWorkflows[event.project_key].lastIndexOf(event.workflow_name) === -1) {
+                                this.unfilteredWorkflows[event.project_key].push(event.workflow_name);
                             }
                         });
 
                         this.currentItem = this.events.length;
 
-                        this.projects = Object.keys(this.groupedEvents).sort();
+                        this.projects = Object.keys(this.unfilteredGroupedEvents).sort();
+                        this.filterEvents();
                     });
             }
         });
     }
 
-    selectTab(t: string): void {
-        this.selectedTab = t;
-    }
+    /**
+     * Make a copy fetched datas and filter on it.
+     * This keep all events to let search them on demand.
+     */
+    filterEvents() {
+        this.workflows = HomeHeatmapComponent.clone(this.unfilteredWorkflows);
+        this.projects = Object.keys(this.unfilteredGroupedEvents).sort();
+        this.groupedEvents = HomeHeatmapComponent.clone(this.unfilteredGroupedEvents);
+        if (this.heatmapSearch) {
+            // filter projects list
+            if (this.heatmapSearch.projects && this.heatmapSearch.projects.length > 0) {
+                this.projects = Object.keys(this.groupedEvents).filter((p) => {
+                    return this.heatmapSearch.projects.indexOf(p) !== -1;
+                }).sort();
+                const projectsToFilter = Object.keys(this.unfilteredWorkflows).filter(proj => this.projects.indexOf(proj) === -1);
+                projectsToFilter.forEach(proj => {
+                    delete this.workflows[proj];
+                });
+            }
+
+            if (this.heatmapSearch.searchCriteria) {
+                const projectsToFilter = new Array<string>();
+                // filter events non matched
+                Object.keys(this.groupedEvents).forEach((project) => {
+                    let projectLength = 0;
+                    if (this.groupedEvents[project]) {
+                        Object.keys(this.groupedEvents[project]).forEach((workflow) => {
+                            if (this.groupedEvents[project][workflow]) {
+                                this.groupedEvents[project][workflow] = this.groupedEvents[project][workflow].filter(event => {
+                                    return event.workflow_name
+                                        .toLowerCase()
+                                        .indexOf(this.heatmapSearch.searchCriteria.toLowerCase()) !== -1 ||
```

Comment:
Can you make a lowercase in a local variable outside of loop to no do the same compute each time ?